### PR TITLE
Add OAuth support: GitHub & Google

### PR DESCRIPTION
**Pull request by jim3ma**
Add multiple OAuth support.

Currently support GitHub & Google.

We can add more OAuth support just extend with the code:

```python
github_blueprint = make_github_blueprint(
    client_id=config.config_github_oauth_client_id,
    client_secret=config.config_github_oauth_client_secret,
    redirect_to="github_login",)

app.register_blueprint(github_blueprint, url_prefix='/login')

github_blueprint.backend = OAuthBackend(ub.OAuth, ub.session, user=current_user, user_required=True)
register_oauth_blueprint(github_blueprint)

@oauth_authorized.connect_via(github_blueprint)
def github_logged_in(blueprint, token):
    if not token:
        flash("Failed to log in with GitHub.", category="error")
        return False

    resp = blueprint.session.get("/user")
    if not resp.ok:
        msg = "Failed to fetch user info from GitHub."
        flash(msg, category="error")
        return False

    github_info = resp.json()
    github_user_id = str(github_info["id"])
    return oauth_update_token(blueprint, token, github_user_id)


# notify on OAuth provider error
@oauth_error.connect_via(github_blueprint)
def github_error(blueprint, error, error_description=None, error_uri=None):
    msg = (
        "OAuth error from {name}! "
        "error={error} description={description} uri={uri}"
    ).format(
        name=blueprint.name,
        error=error,
        description=error_description,
        uri=error_uri,
    )
    flash(msg, category="error")

@app.route('/github')
@github_oauth_required
def github_login():
    if not github.authorized:
        return redirect(url_for('github.login'))
    account_info = github.get('/user')
    if account_info.ok:
        account_info_json = account_info.json()
        return bind_oauth_or_register(github_blueprint.name, account_info_json['id'], 'github.login')
    flash(_(u"GitHub Oauth error, please retry later."), category="error")
    return redirect(url_for('login'))
```

### CODE REVIEW

**Comment by jim3ma** [comment]
Todo:

* Link/unlink OAuth

File: cps/web.py [hunk 1]
```diff
@@ -3883,3 +3944,214 @@ def convert_bookformat(book_id):
     else:
         flash(_(u"There was an error converting this book: %(res)s", res=rtn), category="error")
     return redirect(request.environ["HTTP_REFERER"])
+
+
+def register_oauth_blueprint(blueprint):
```
Comment (bodybybuddha) [inline cps/web.py]: Now that we are using blueprints - could we try to move the new features into their own modules?  Just trying to minimize the amount of stuff we keep putting into web.py.

File: cps/templates/login.html [hunk 2]
```diff
@@ -21,6 +21,17 @@ <h2 style="margin-top: 0">{{_('Login')}}</h2>
     {% if remote_login %}
     <a href="{{url_for('remote_login')}}" class="pull-right">{{_('Log in with magic link')}}</a>
     {% endif %}
+    <a href="{{url_for('github_login')}}" class="pull-right">
```
Comment (bodybybuddha) [inline cps/templates/login.html]: We may want to pass in a variable to this page to determine if the GitHub and/or Google icons should be shown for login.  Actually, we may want to pass in a general OAuth configured variable to determine if any additional icons be shown and then individual flags for each service?

Comment (jim3ma) [reply cps/templates/login.html]: We can use the config

```
config.config_use_github_oauth
```

and

```
config.config_use_google_oauth
```

File: cps/web.py (hunk 1 above)
Comment (jim3ma) [reply cps/web.py]: This function is just record which oauth is used for calibre web.
We will check for oauth settings in sessions, we need know how many oauth is enabled in calibre web.

**Comment by halkeye** [comment]
Oooh, this is super exciting.

I was thinking of adding a header (and I still might) to do oauth support on nginx, but this is way lazier for me if you get it done.

PS, you have conflicted files right now.

File: cps/templates/config_edit.html [hunk 3]
```diff
@@ -162,6 +162,36 @@ <h4 class="panel-title">
       </div>
     </div>
     {% endif %}
+    <div class="form-group">
+      <input type="checkbox" id="config_use_github_oauth" name="config_use_github_oauth" data-control="github-oauth-settings" {% if content.config_use_github_oauth %}checked{% endif %}>
+      <label for="config_use_github_oauth">{{_('Use')}} GitHub OAuth</label>
+      <a href="https://github.com/settings/developers" target="_blank" style="margin-left: 5px">{{_('Obtain GitHub OAuth Credentail')}}</a>
```
Comment (bitstein) [inline cps/templates/config_edit.html]: `Credentail` should be `Credential` in both instances.

```suggestion
      <a href="https://github.com/settings/developers" target="_blank" style="margin-left: 5px">{{_('Obtain GitHub OAuth Credential')}}</a>
```

**Comment by bkanuka** [comment]
@jim3ma Are you able to fix the conflicts so that this can merge? I'm excited about the prospect of OAuth support!

**Comment by OzzieIsaacs** [comment]
You could check out the development branch, it‘s merged there, and it should work.(for testing purposes only)

**Comment by rbmaster** [comment]
What about Plex OAuth? It would be nice being able to sign in using the plex account, Calibre-web really goes nicely on a media ecosystem (Emby, Plex, Tautulli, Ombi, Organizr, etc etc).

**Comment by xfouloux** [comment]
Can plex be added as a oauth provider like Tautulli do ?

https://github.com/Tautulli/Tautulli

Thanks !

**Pull request merged**